### PR TITLE
(chaoslib/litmus): Added service kill lib

### LINKED ISSUE #809: Kill Service experiment
FEATURE REQUEST

**What happened**:

Similar to kill pod, have a experiment to kill service. 
**What you expected to happen**:

The kill service experiment will remove the service associated with the pod or disrupt the type of service to observe the effect of how the environment would respond if the service were to be killed. 
**How to reproduce it (as minimally and precisely as possible)**:

New feature request 
**Anything else we need to know?**:

The kill service would help in deeper understanding of application deployment as well as add a new feature set. 

I would be happy to work on this issue.

### LINKED ISSUE #827: Icon for Container Kill experiment
Design icons to show case the following experiments. Each experiment should have an icon in
https://hub.litmuschaos.io/charts/generic
This issue is to track the icon for **Container Kill** experiment

### PR DESCRIPTION
**Pull request by kmjayadeep**
Signed-off-by: Jayadeep KM <kmjayadeep@gmail.com>

**What this PR does / why we need it**:

Adds a chaoslib for deleting and reviving services. It is useful to monitor the behavior of microservices when connection to other services / databases are disabled for a certain period.

fixes #809 

This is a work in progress PR. I will post logs and add README once the approach is finalized.
Here is the yaml file I use to test this lib

```yaml
---
- hosts: localhost
  connection: local
  gather_facts: no

  vars:
    c_experiment: "service-delete"
    s_ns: "default"
    s_name: "mongodb-deployment"
    action: "revive"

  tasks:
    - include_tasks: ../litmus/chaoslib/litmus/service_delete_by_litmus.yml      

```

`s_name` refers to the service name
`action` can be `delete` or `revive`. While writing the experiment, we can call this lib with `action` as `delete` and do a liveness test of the app. After the experiment, the service can be revived with `revive` action.

Here is the explanation of ansible steps

### Action = delete

- Get the service info and print the details as debug message
- Get the service description in yaml format
- Save the yaml to a local temporary file
- Save the yaml as a `ConfigMap` with the same name as service under the same namespace. This is required to revive the service later once the experiment is finished

### Action = revive

- Get the yaml file from `ConfigMap` from the delete action
- Copy yaml to a temporary location
- Create service from the provided yaml from temporary location
- Delete `ConfigMap` for cleanup

**Checklist**

* [x] Does this PR have a corresponding GitHub issue?
* [ ] Have you included relevant README for the chaoslib/experiment with details?
* [ ] Have you added debug messages where necessary? 
* [ ] Have you added task comments where necessary? 
* [ ] Have you tested the changes for possible failure conditions?
* [ ] Have you provided the positive & negative test logs for the litmusbook execution?
* [ ] Does the litmusbook ensure idempotency of cluster state?, i.e., is cluster restored to original state?
* [ ] Have you used non-shell/command modules for Kubernetes tasks?
* [ ] Have you (jinja) templatized custom scripts that is run by the litmusbook, if any? 
* [ ] Have you (jinja) templatized Kubernetes deployment manifests used by the litmusbook, if any?
* [ ] Have you reused/created util functions instead of repeating tasks in the litmusbook?
* [ ] Do the artifacts follow the appropriate directory structure? 
* [ ] Have you isolated storage (eg: OpenEBS) specific implementations, checks? 
* [ ] Have you isolated platform (eg: baremetal kubeadm/openshift/aws/gcloud) specific implementations, checks?  
* [ ] Are the ansible facts well defined? Is the scope explicitly set for playbook & included utils?
* [ ] Have you ensured minimum/careful usage of shell utilities (awk, grep, sed, cut, xargs etc.,)?
* [ ] Can the limtusbook be executed both from within & outside a container (configurable paths, no hardcode)?
* [ ] Can you suggest the minimal resource requirements for the litmusbook execution?
* [ ] Does the litmusbook job artifact carry comments/default options/range for the ENV tunables?
* [ ] Has the litmusbooks been linted? 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by kmjayadeep** [comment]
@ksatchit Could you please confirm whether the approach is correct? My goal is to write a test scenario for a sample application which uses mongodb. I think it would be better to create separate PRs for lib, experiment and chart examples. What do you suggest?

**Comment by ksatchit** [comment]
Approach looks fine @kmjayadeep. And agree in the separate PRs as well!

**Comment by kmjayadeep** [comment]
@ksatchit Here are the logs
## action=delete
yaml file:
```yaml
---
- hosts: localhost
  connection: local
  gather_facts: no

  vars:
    c_experiment: "service-delete"
    s_ns: "default"
    s_name: "mongodb-deployment"
    action: "delete"

  tasks:
    - include_tasks: ../litmus/chaoslib/litmus/service_delete_by_litmus.yml
```
Ansible logs
```log
PLAY [localhost] **********************************************************************************

TASK [include_tasks] ******************************************************************************
included: /Users/jayadeep/workspace/opensource/litmus/chaoslib/litmus/service_delete_by_litmus.yml for localhost

TASK [Delete Service] *****************************************************************************
included: /Users/jayadeep/workspace/opensource/litmus-test/../litmus/chaoslib/litmus/service_delete.yml for localhost

TASK [Get the service] ****************************************************************************
ok: [localhost]

TASK [debug] **************************************************************************************
ok: [localhost] => {
    "msg": "Deleting service {'changed': False, 'resources': [{'kind': 'Service', 'apiVersion': 'v1', 'metadata': {'name': 'mongodb-deployment', 'namespace': 'default', 'selfLink': '/api/v1/namespaces/default/services/mongodb-deployment', 'uid': '1276d30e-edcd-11e9-9f0a-025000000001', 'resourceVersion': '487007', 'creationTimestamp': '2019-10-13T15:21:04Z', 'labels': {'app': 'mongodb'}, 'annotations': {'kubectl.kubernetes.io/last-applied-configuration': '{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{},\"creationTimestamp\":\"2019-10-13T15:21:04Z\",\"labels\":{\"app\":\"mongodb\"},\"name\":\"mongodb-deployment\",\"namespace\":\"default\",\"resourceVersion\":\"287153\",\"selfLink\":\"/api/v1/namespaces/default/services/mongodb-deployment\",\"uid\":\"1276d30e-edcd-11e9-9f0a-025000000001\"},\"spec\":{\"clusterIP\":\"10.100.224.52\",\"externalTrafficPolicy\":\"Cluster\",\"ports\":[{\"nodePort\":30006,\"port\":27017,\"protocol\":\"TCP\",\"targetPort\":27017}],\"selector\":{\"app\":\"mongodb\"},\"sessionAffinity\":\"None\",\"type\":\"NodePort\"},\"status\":{\"loadBalancer\":{\"ingress\":[{\"hostname\":\"localhost\"}]}}}\\n'}}, 'spec': {'ports': [{'protocol': 'TCP', 'port': 27017, 'targetPort': 27017, 'nodePort': 30006}], 'selector': {'app': 'mongodb'}, 'clusterIP': '10.100.224.52', 'type': 'NodePort', 'sessionAffinity': 'None', 'externalTrafficPolicy': 'Cluster'}, 'status': {'loadBalancer': {'ingress': [{'hostname': 'localhost'}]}}}], 'failed': False}"
}

TASK [Saving service yaml for recreation of service after deletion] *******************************
changed: [localhost]

TASK [Copy service yaml to file] ******************************************************************
changed: [localhost -> localhost]

TASK [Create ConfigMap to save the yaml file] *****************************************************
changed: [localhost]

TASK [Delete service] *****************************************************************************
changed: [localhost]

TASK [Revive Service] *****************************************************************************
skipping: [localhost]

PLAY RECAP ****************************************************************************************
localhost                  : ok=8    changed=4    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
```

##action=revive
yaml file: same as above with action=revive
log:
```

PLAY [localhost] **********************************************************************************

TASK [include_tasks] ******************************************************************************
included: /Users/jayadeep/workspace/opensource/litmus/chaoslib/litmus/service_delete_by_litmus.yml for localhost

TASK [Delete Service] *****************************************************************************
skipping: [localhost]

TASK [Revive Service] *****************************************************************************
included: /Users/jayadeep/workspace/opensource/litmus-test/../litmus/chaoslib/litmus/service_revive.yml for localhost

TASK [Get ConfigMap data] *************************************************************************
changed: [localhost]

TASK [Copy service yaml to file] ******************************************************************
ok: [localhost -> localhost]

TASK [Recreate the Service object] ****************************************************************
changed: [localhost]

TASK [Delete ConfigMap] ***************************************************************************
changed: [localhost]

PLAY RECAP ****************************************************************************************
localhost                  : ok=6    changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
```

File: chaoslib/litmus/service_delete.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  args:
+    executable: /bin/bash
+  register: service_yaml
+
+- name: Copy service yaml to file
+  local_action: copy content="{{ service_yaml.stdout }}" dest="/tmp/service.yaml"
+
+- name: Create ConfigMap to save the yaml file
+  shell: |
+    kubectl create configmap -n "{{ s_ns }}" "{{s_name}}" --from-file=/tmp/service.yaml
```
Comment (ksatchit) [inline chaoslib/litmus/service_delete.yml]: :+1:

File: chaoslib/litmus/service_delete.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+- name: Get the service
+  k8s_facts:
+    kind: Service
+    namespace: "{{ s_ns }}"
```
Comment (ksatchit) [inline chaoslib/litmus/service_delete.yml]: A very minor comment, open for feedback.  

Across the libs, we are looking at keeping the var names consistent. `{{ a_ns }}` is what is being used to filter the app's namespace via the standard vars step: `a_ns: "{{ lookup('env','APP_NAMESPACE') }}"`. 
With this, only the vars specific to that chaoslib are being passed explicitly via include vars.  

The service namespace is essentially the same as the app namespace. Can we reuse that? There could be cases where service kills are effected at one ns, but a check is conducted against an app (like a liveness client) in another namespace, in which case we could extend the convention to `<type>_a_ns`.

Comment (kmjayadeep) [reply chaoslib/litmus/service_delete.yml]: I was thinking that `s_ns` would be an acronym for service namespace. But as you said, i will rename it to `a_ns` to keep it consistent with other experiments.

Comment (kmjayadeep) [reply chaoslib/litmus/service_delete.yml]: So what about `s_name` ? Do you think it should be `a_name` ?

Comment (ksatchit) [reply chaoslib/litmus/service_delete.yml]: `s_name` should still be fine @kmjayadeep ! It allows for the possibility of an experiment job to take the service name of the app as an ENV along w/ the regular info (label, ns)

**Comment by ksatchit** [comment]
Would you like to add anything else on this PR @kmjayadeep - if not we could remove the WIP status!

**Comment by kmjayadeep** [comment]
> Would you like to add anything else on this PR @kmjayadeep - if not we could remove the WIP status!

Done

**Pull request merged**